### PR TITLE
:sparkles: Enable gitlab Packaging Reporting

### LINKED ISSUE #2940: Feature: Enable gitlab Packaging Reporting
**Is your feature request related to a problem? Please describe.**
As a gitlab project owner or consumer, I'd like to get scoring on the `Packaging Check` of those projects housed in gitlab.

**Describe the solution you'd like**
Detect Packaging commands within the CI/CD pipelines that package and push artifacts.

**Additional context**
Gitlab projects are being added to scorecard on an experimental basis, but some of the individual checks are either not implemented at all, or in an incomplete status.

### PR DESCRIPTION
**Pull request by jimrobison**
#### What kind of change does this PR introduce?

Feature to accomplish initial scoring for gitlab package check.

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?
- Currently there isn't a score at all for gitlab projects with regards to `Packaging`

#### What is the new behavior (if this is a feature change)?**

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes
Fixes #2940 

#### Special notes for your reviewer
- The modifications take advantage of a gitlab API that combines the CI/CD yaml files into a single yaml file that makes it easier to confirm/query for actions taking place.
- Currently, what is searched for:
   - `docker push`
   - `nuget push`
   - `poetry publish`
   - `twine upload`

#### Does this PR introduce a user-facing change?

Implementation of score on Packaging check for gitlab projects.

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
NONE
```

### CODE REVIEW

File: clients/gitlabrepo/tarball.go [hunk 1]
```diff
@@ -114,40 +122,85 @@ func (handler *tarballHandler) setup() error {
...
+	handler.files = append(handler.files,
+		strings.TrimPrefix(ciYaml.Name(), filepath.Clean(handler.tempDir)+string(os.PathSeparator)))
+	return nil
+}
+
+func (handler *tarballHandler) apiFunction(url, tempDir string, repoFile *os.File) error {
 	req, err := http.NewRequestWithContext(handler.ctx, http.MethodGet, url, nil)
 	if err != nil {
-		return fmt.Errorf("http.NewRequestWithContext: %w", err)
+		return fmt.Errorf("%w io.Copy: %v", errTarballNotFound, err)
```
Comment (raghavkaul) [inline clients/gitlabrepo/tarball.go:186]: Isn't this still an `http.NewRequestWithContext` error?

**Review by raghavkaul — COMMENTED** [review]
Thanks for the changes!

File: checks/fileparser/gitlab_workflow.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fileparser
+
+// IsGitlabWorkflowFile determines if a file is a workflow
+// as a callback to use for repo client's ListFiles() API.
+func IsGitlabWorkflowFile(pathfn string) (bool, error) {
+	return pathfn == "gitlabscorecard_flattened_ci.yaml", nil
```
[the remark this answers is not in the document]
Comment (jimrobison) [reply checks/fileparser/gitlab_workflow.go:20]: This is a file that is generated by the gitlab API that has the flattened ci/cd contents as a single entry. Remote/Locals/Templates are all yanked in to make a complete picture of the ci/cd jobs.

File: clients/gitlabrepo/tarball.go (hunk 1 above)
Comment (jimrobison) [reply clients/gitlabrepo/tarball.go:186]: Yep, I'll get that reverted... Refactoring, copy/paste issue.

Comment (jimrobison) [reply clients/gitlabrepo/tarball.go]: Just pushed the commit, reverting this. Thanks for the catch.

File: clients/gitlabrepo/tarball.go [hunk 3]
```diff
@@ -114,40 +122,85 @@ func (handler *tarballHandler) setup() error {
...
+	if err != nil {
+		return fmt.Errorf("os.ReadFile: %w", err)
+	}
+	var result gitLabLint
+	err = json.Unmarshal(byteValue, &result)
+	if err != nil {
+		return fmt.Errorf("json.Unmarshal: %w", err)
+	}
+
+	ciYaml, err := os.Create(tempDir + "/gitlabscorecard_flattened_ci.yaml")
```
Comment (raghavkaul) [inline clients/gitlabrepo/tarball.go:161]: Where is the flattening happening? Does L147 give us a single file or multiple files?

Comment (jimrobison) [reply clients/gitlabrepo/tarball.go:161]: * [Pulling of the raw linted file](https://github.com/kgangerlm/scorecard-gitlab/blob/gitlab-packaging/clients/gitlabrepo/tarball.go#L141-L147)
* [Extracting the merged yaml (flattened file)](https://github.com/kgangerlm/scorecard-gitlab/blob/gitlab-packaging/clients/gitlabrepo/tarball.go#L151-L166) and saving to the file.

gitlab documentation: https://docs.gitlab.com/ee/api/lint.html

Comment (jimrobison) [reply clients/gitlabrepo/tarball.go:161]: Just noticed what you were asking... There is a json result that contains a few related CI/CD linting elements/attributes. The field that I care about is the `merged_yaml`. This is being saved off as a yaml file (hardcoded name).

Comment (raghavkaul) [reply clients/gitlabrepo/tarball.go:161]: Gotcha, thanks!

**Review by raghavkaul — APPROVED** [review]

**Comment by raghavkaul** [comment]
@jimrobison Could you also update README.md that the Packaging check is supported?

**Comment by jimrobison** [comment]
> @jimrobison Could you also update README.md that the Packaging check is supported?

Are you referring to this [README](https://github.com/ossf/scorecard/blob/main/README.md?plain=1#L450)? It already has it annotated that it is `Fully Supported`.

**Comment by raghavkaul** [comment]
Ah, gotcha. That sounds like it wasn't true before this PR (would be good to review other checks as well, I might have glossed over them). But we should be good to merge.

**Pull request merged**